### PR TITLE
Backports 0.15 pr23

### PR DESCRIPTION
**Pull request by PastaPastaPasta**

### CODE REVIEW

File: src/Makefile.test.include [hunk 1]
```diff
@@ -109,12 +109,13 @@ endif
...
 test_test_dash_SOURCES = $(BITCOIN_TESTS) $(JSON_TEST_FILES) $(RAW_TEST_FILES)
 test_test_dash_CPPFLAGS = $(AM_CPPFLAGS) $(BITCOIN_INCLUDES) -I$(builddir)/test/ $(TESTDEFS) $(EVENT_CFLAGS)
-test_test_dash_LDADD = $(LIBBITCOIN_SERVER) $(LIBBITCOIN_CLI) $(LIBBITCOIN_COMMON) $(LIBBITCOIN_UTIL) $(LIBBITCOIN_CONSENSUS) $(LIBBITCOIN_CRYPTO) $(LIBUNIVALUE) \
-  $(LIBLEVELDB) $(LIBLEVELDB_SSE42) $(LIBMEMENV) $(BACKTRACE_LIB) $(BOOST_LIBS) $(BOOST_UNIT_TEST_FRAMEWORK_LIB) $(LIBSECP256K1) $(EVENT_LIBS) $(EVENT_PTHREADS_LIBS)
-test_test_dash_CXXFLAGS = $(AM_CXXFLAGS) $(PIE_FLAGS)
+test_test_dash_LDADD =
 if ENABLE_WALLET
 test_test_dash_LDADD += $(LIBBITCOIN_WALLET)
 endif
+test_test_bitcoin_LDADD += $(LIBBITCOIN_SERVER) $(LIBBITCOIN_CLI) $(LIBBITCOIN_COMMON) $(LIBBITCOIN_UTIL) $(LIBBITCOIN_CONSENSUS) $(LIBBITCOIN_CRYPTO) $(LIBUNIVALUE) \
```
Comment (UdjinM6) [inline src/Makefile.test.include]:
```suggestion
test_test_ashd_LDADD += $(LIBBITCOIN_SERVER) $(LIBBITCOIN_CLI) $(LIBBITCOIN_COMMON) $(LIBBITCOIN_UTIL) $(LIBBITCOIN_CONSENSUS) $(LIBBITCOIN_CRYPTO) $(LIBUNIVALUE) \
```

File: src/Makefile.test.include [hunk 2]
```diff
@@ -109,12 +109,13 @@ endif
...
+  $(LIBLEVELDB) $(LIBLEVELDB_SSE42) $(LIBMEMENV) $(BOOST_LIBS) $(BOOST_UNIT_TEST_FRAMEWORK_LIB) $(LIBSECP256K1) $(EVENT_LIBS) $(EVENT_PTHREADS_LIBS)
+test_test_bitcoin_CXXFLAGS = $(AM_CXXFLAGS) $(PIE_FLAGS)
```
Comment (UdjinM6) [inline src/Makefile.test.include]:
```suggestion
test_test_dash_CXXFLAGS = $(AM_CXXFLAGS) $(PIE_FLAGS)
```

File: src/Makefile.test.include [hunk 3]
```diff
@@ -109,12 +109,13 @@ endif
...
 test_test_dash_CPPFLAGS = $(AM_CPPFLAGS) $(BITCOIN_INCLUDES) -I$(builddir)/test/ $(TESTDEFS) $(EVENT_CFLAGS)
-test_test_dash_LDADD = $(LIBBITCOIN_SERVER) $(LIBBITCOIN_CLI) $(LIBBITCOIN_COMMON) $(LIBBITCOIN_UTIL) $(LIBBITCOIN_CONSENSUS) $(LIBBITCOIN_CRYPTO) $(LIBUNIVALUE) \
-  $(LIBLEVELDB) $(LIBLEVELDB_SSE42) $(LIBMEMENV) $(BACKTRACE_LIB) $(BOOST_LIBS) $(BOOST_UNIT_TEST_FRAMEWORK_LIB) $(LIBSECP256K1) $(EVENT_LIBS) $(EVENT_PTHREADS_LIBS)
-test_test_dash_CXXFLAGS = $(AM_CXXFLAGS) $(PIE_FLAGS)
+test_test_dash_LDADD =
 if ENABLE_WALLET
 test_test_dash_LDADD += $(LIBBITCOIN_WALLET)
 endif
+test_test_bitcoin_LDADD += $(LIBBITCOIN_SERVER) $(LIBBITCOIN_CLI) $(LIBBITCOIN_COMMON) $(LIBBITCOIN_UTIL) $(LIBBITCOIN_CONSENSUS) $(LIBBITCOIN_CRYPTO) $(LIBUNIVALUE) \
+  $(LIBLEVELDB) $(LIBLEVELDB_SSE42) $(LIBMEMENV) $(BOOST_LIBS) $(BOOST_UNIT_TEST_FRAMEWORK_LIB) $(LIBSECP256K1) $(EVENT_LIBS) $(EVENT_PTHREADS_LIBS)
```
Comment (UdjinM6) [inline src/Makefile.test.include]:
```suggestion
  $(LIBLEVELDB) $(LIBLEVELDB_SSE42) $(LIBMEMENV) $(BACKTRACE_LIB) $(BOOST_LIBS) $(BOOST_UNIT_TEST_FRAMEWORK_LIB) $(LIBSECP256K1) $(EVENT_LIBS) $(EVENT_PTHREADS_LIBS)
```

File: test/functional/multiwallet.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    def run_test(self):
+        w1 = self.nodes[0] / "wallet/w1"
+        w1.generate(1)
+
+        #accessing wallet RPC without using wallet endpoint fails
+        assert_raises_jsonrpc(-32601, "Method not found", self.nodes[0].getwalletinfo)
+
+        #check w1 wallet balance
+        walletinfo = w1.getwalletinfo()
+        assert_equal(walletinfo['immature_balance'], 50)
```
Comment (UdjinM6) [inline test/functional/multiwallet.py]:
```suggestion
        assert_equal(walletinfo['immature_balance'], 500)
```

File: test/functional/multiwallet.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+        #check w1 wallet balance
+        w2 = self.nodes[0] / "wallet/w2"
+        walletinfo = w2.getwalletinfo()
+        assert_equal(walletinfo['immature_balance'], 0)
+
+        w3 = self.nodes[0] / "wallet/w3"
+        
+        w1.generate(101)
+        assert_equal(w1.getbalance(), 100)
```
Comment (UdjinM6) [inline test/functional/multiwallet.py]:
```suggestion
        assert_equal(w1.getbalance(), 1000)
```

File: src/Makefile.test.include (hunk 1 above)
Comment (PastaPastaPasta) [reply src/Makefile.test.include]: Umm. You sure about that?

Comment (PastaPastaPasta) [reply src/Makefile.test.include]:
```suggestion
test_test_dash_LDADD += $(LIBBITCOIN_SERVER) $(LIBBITCOIN_CLI) $(LIBBITCOIN_COMMON) $(LIBBITCOIN_UTIL) $(LIBBITCOIN_CONSENSUS) $(LIBBITCOIN_CRYPTO) $(LIBUNIVALUE) \
```

Comment (UdjinM6) [reply src/Makefile.test.include]: ooops, typo 😊 should be `dash` ofc :)

**Comment by UdjinM6** [comment]
Guess we also have to finish 10297 backport to fix `DisconnectBlock` issue https://travis-ci.org/dashpay/dash/jobs/567341959#L2131. We also need 10995 and 11002 to fix `resendwallettransactions` crash and 10683 too fix `generate` in `multiwallet.py`. See https://github.com/UdjinM6/dash/commits/pr3043, Travis: https://travis-ci.org/UdjinM6/dash/builds/567602297.

**Comment by PastaPastaPasta** [comment]
I believe #10683 will need more changes due to our changes that mining isn't compiled in by default

**Comment by PastaPastaPasta** [comment]
Compiles and passes tests locally

**Comment by PastaPastaPasta** [comment]
@codablock please review

File: src/validation.cpp [hunk 6]
```diff
@@ -1387,7 +1387,7 @@ int ApplyTxInUndo(Coin&& undo, CCoinsViewCache& view, const COutPoint& out)
 
 /** Undo the effects of this block (with given index) on the UTXO set represented by coins.
  *  When FAILED is returned, view is left in an indeterminate state. */
-static DisconnectResult DisconnectBlock(const CBlock& block, CValidationState& state, const CBlockIndex* pindex, CCoinsViewCache& view)
+static DisconnectResult DisconnectBlock(const CBlock& block, const CBlockIndex* pindex, CCoinsViewCache& view)
```
Comment (UdjinM6) [inline src/validation.cpp]: this belongs to "drop state..." commit

File: src/wallet/rpcwallet.cpp [hunk 7]
```diff
@@ -2948,6 +2948,7 @@ UniValue setbip69enabled(const JSONRPCRequest& request)
     return NullUniValue;
 }
 
+#if ENABLE_MINER
```
Comment (UdjinM6) [inline src/wallet/rpcwallet.cpp]: would prefer to wrap `#include "rpc/mining.h"` too https://github.com/UdjinM6/dash/commit/3567fdc79d41c6dab3c75aa7ece2d5a7a8543d34#diff-df7d84ff2f53fcb2a0dc15a3a51e55ceR17

Comment (PastaPastaPasta) [reply src/wallet/rpcwallet.cpp]: https://github.com/dashpay/dash/pull/3043/commits/42cf74dec48e81a528b181e199a13e241aa2aacc

File: src/validation.cpp (hunk 6 above)
Comment (PastaPastaPasta) [reply src/validation.cpp]: https://github.com/dashpay/dash/pull/3043/commits/f66ff6be02da31c551fccbd132ed7f45420eab7e

**Comment by UdjinM6** [comment]
Travis is having its usual issues again it seems...

File: src/qt/forms/sendcoinsdialog.ui [hunk 8]
```diff
@@ -762,7 +762,7 @@
           <item>
            <widget class="QLabel" name="fallbackFeeWarningLabel">
             <property name="toolTip">
-             <string>Using the fallbackfee can result in sending a transaction that will take several hours or days (or never) to confirm. Consider choosing your fee manually or wait until your have validated the complete chain.</string>
+             <string>Using the fallbackfee can result in sending a transaction that will take several hours or days (or never) to confirm. Consider choosing your fee manually or wait until you have validated the complete chain.</string>
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply src/qt/forms/sendcoinsdialog.ui]: Isn't that pulled automatically?

[the remark this answers is not in the document]
Comment (codablock) [reply src/qt/forms/sendcoinsdialog.ui]: don't know to be honest as I've never gone through the translation process. @UdjinM6 can answer this. If it's done automatically, then utACK

[the remark this answers is not in the document]
Comment (UdjinM6) [reply src/qt/forms/sendcoinsdialog.ui]: yeah, ts files are generated automagically via `make translate` https://github.com/dashpay/dash/blob/master/doc/translation_process.md#writing-code-with-translations and we usually batch them together in one single `Translations <some_date>` PR.

[the remark this answers is not in the document]
Comment (UdjinM6) [reply src/qt/forms/sendcoinsdialog.ui]: I mean, you could follow the original bitcoin PR and do this here but strings are going to be broken anyway (other languages won't recognize new `en` string and won't be able to map translations for it properly). Either way is fine for me.

[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply src/qt/forms/sendcoinsdialog.ui]: I prefer to keep it as is

**Pull request merged**